controllers/liveupdate_controller.go

```diff
@@ -0,0 +1,653 @@
+	if err := r.reconcileNewConfigMap(); err != nil {
+		return reconcile.Result{}, err
+	}
+	// Reconcile ping service with new DNS query
+	if err := r.reconcileNewPingService(); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Obtain existing statefulSet
+	namespace := r.infinispan.Namespace
+	statefulSetName := r.infinispan.GetStatefulSetName()
+	sourceStatefulSet := &appsv1.StatefulSet{}
+	err := r.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: statefulSetName}, sourceStatefulSet)
+
+	if err != nil && errors.IsNotFound(err) {
+		r.reqLogger.Error(err, "Could not find existing statefulSet'%s'", statefulSetName)
+		return reconcile.Result{}, err
+	}
+
+	// Create a new statefulSet based on the existing one
+	targetStatefulSet := &appsv1.StatefulSet{}
+	err = r.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: r.targetName}, targetStatefulSet)
+	if err != nil && errors.IsNotFound(err) {
+		targetStatefulset := sourceStatefulSet.DeepCopy()
+		volumes := targetStatefulset.Spec.Template.Spec.Volumes
+		// Change the config name
+		for _, volume := range volumes {
+			if volume.Name == ConfigVolumeName {
+				volume.ConfigMap.Name = r.newConfigName
+			}
+		}
+		targetStatefulset.Name = r.targetName
+		targetStatefulset.ObjectMeta.ResourceVersion = ""
+		// Update the image references
+		container := &targetStatefulset.Spec.Template.Spec.Containers[0]
+		container.Image = consts.DefaultImageName
+		env := container.Env
+		for o, envVar := range env {
+			if envVar.Name == "DEFAULT_IMAGE" {
+				a := corev1.EnvVar{
+					Name:      envVar.Name,
+					Value:     consts.DefaultImageName,
+					ValueFrom: envVar.ValueFrom,
+				}
+				env[o] = a
+				break
+			}
+		}
+		container.Env = env
+
+		err = r.Create(r.ctx, targetStatefulset)
+		if err != nil {
+			r.reqLogger.Error(err, "failed to create new statefulSet '%s'", r.targetName)
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{RequeueAfter: consts.DefaultWaitClusterNotWellFormed}, err
+	}
+	// If the new statefulSet was already created, check if the pods are ready
+	podList, err := r.podList(targetStatefulSet)
+	if err != nil {
+		r.reqLogger.Error(err, "Error getting pods of the new statefulSet '%s'", r.targetName)
+		return reconcile.Result{}, err
+	}
+
+	for _, pod := range podList.Items {
+		if !kube.IsPodReady(pod) {
+			r.reqLogger.Info(fmt.Sprintf("Pod '%s' not ready", pod.Name))
+			return reconcile.Result{RequeueAfter: consts.DefaultWaitOnCluster}, err
+		}
+		// Add a "version" label to pods
+		_, err = controllerutil.CreateOrUpdate(r.ctx, r.Client, &pod, func() error {
+			if pod.CreationTimestamp.IsZero() {
+				return errors.NewNotFound(corev1.Resource(""), pod.Name)
+			}
+			labels := pod.ObjectMeta.Labels
+			labels["version"] = strings.Split(consts.DefaultImageName, ":")[1]
+			pod.Labels = labels
+			return nil
+		})
+	}
+
+	// Move to the next stage
+	return r.updateStage(ispnv1.LiveUpgradeStart)
```

Comment:
We never check that the number of ready pods is equal to the number of replicas in the StatefulSet spec. As the StatefulSet spec brings up the cluster a pod at a time, I think if the timing is wrong it's possible for replica = 2, but `len(podList.Items) == 1`, resulting in the controller progressing to `LiveUpgradeStart` prematurely.